fenix/app/src/main/java/org/mozilla/fenix/messaging/state/MessagingMiddleware.kt

```diff
@@ -69,15 +70,20 @@ class MessagingMiddleware(
         oldMessage: Message,
         context: AppStoreMiddlewareContext,
     ) {
+        val newMessage = runBlocking {
```

Comment:
how come the addition of runBlocking? is it safe to thread-block here?

Would this be an ok alternative?

```
val newMessage
val newMessages
coroutineScope.launch {
            newMessage = controller.onMessageDisplayed(oldMessage)
            newMessages = if (!newMessage.isExpired) {
                updateMessage(context, oldMessage, newMessage)
            } else {
                removeMessage(context, oldMessage)
            }
        }

if (newMessage.isExpired) {
   consumeMessageToShowIfNeeded(context, oldMessage)
}

context.store.dispatch(UpdateMessages(newMessages))
```

Reply:
Unfortunately no, `newMessage.isExpired` could happen before `newMessage` is set.  This was the original code before Jame's change.   

```
        val newMessage = controller.updateMessageAsDisplayed(oldMessage)
        val newMessages = if (!newMessage.isExpired) {
            updateMessage(context, oldMessage, newMessage)
        } else {
            consumeMessageToShowIfNeeded(context, oldMessage)
            removeMessage(context, oldMessage)
        }
        context.dispatch(UpdateMessages(newMessages))
        coroutineScope.launch {
            controller.onMessageDisplayed(newMessage) 
```
I'm just trying to restore the way it was called.

Reply:
Gotcha gotcha. I'm just having a hard time discerning whether this section can be a simple revert without knowing how the removal of `updateMessageAsDisplayed` can be 1 to 1 replaced with `onMessageDisplayed`

Reply:
Hmm you're right.  Maybe I should just put back what we had before the change.

Reply:
I'll look for a better solution.  This right now is risky.  Thanks @MozillaNoah

Reply:
The alternative solution that is safer is to only dispatch to the store using main thread.  [This](https://searchfox.org/mozilla-mobile/rev/21548505501d6cee5ce2aac3244821269bc7969d/firefox-android/android-components/components/lib/state/src/main/java/mozilla/components/lib/state/Middleware.kt#42) stated that we should only dispatch from the calling thread.  In this case it's the main thread.